<h1 style="text-align: center;">
Intelligent Trading Solution Work Space
</h1>

<h2 style="text-align: left;"> 1. Data Retrival Module </h2>

In [1]:
# Importing Packages
import pandas as pd
import numpy as np
import yfinance as yf
import requests
from config import settings



In [6]:
class DataRetrieval:
    def __init__(self, ticker):
        self.ticker = ticker
        self.apikey = settings.alpha_api_key
        
    def Histrocal_prices(self):
        # inintiatiating the API
        url = "https://www.alphavantage.co/query"
        params = {
            "function": "TIME_SERIES_DAILY",
            "symbol": self.ticker,
            "apikey": self.apikey,
            "outputsize": "full"
        }
        response = requests.get(url=url, params=params)
        return response


        

In [ ]:
import requests


url = "https://www.alphavantage.co/query"
params = {
    "function": "TIME_SERIES_DAILY",
    "symbol": "AAPL",
    "apikey": "your_api_key",
    "outputsize": "full"
}
response = requests.get(url, params=params)
data = response.json()
print(data)


: 

In [3]:
dir(settings)
settings.alpha_api_key
dir(settings)
settings.finnhub_api_key

'cu2fta9r01ql7sc7c0pgcu2fta9r01ql7sc7c0q0'

<h2 style="text-align: left; font-family: 'Arial', sans-serif; color: #4CAF50;">
  🚀 Testing Space: Where Magic Happens! ✨
</h2>


### Data retrival-  Hist data

In [4]:
from data_retrieval import DataRetrieval
# Initializing  it 
ticker = "TSLA"
DR = DataRetrieval(ticker=ticker)

response = DR.Histrocal_prices()

AttributeError: 'DataRetrieval' object has no attribute 'Histrocal_prices'

In [33]:
response.status_code

200

In [3]:
dir(response)

['Close',
 'High',
 'Low',
 'Open',
 'T',
 'Volume',
 '_AXIS_LEN',
 '_AXIS_ORDERS',
 '_AXIS_TO_AXIS_NUMBER',
 '_HANDLED_TYPES',
 '__abs__',
 '__add__',
 '__and__',
 '__annotations__',
 '__array__',
 '__array_priority__',
 '__array_ufunc__',
 '__arrow_c_stream__',
 '__bool__',
 '__class__',
 '__contains__',
 '__copy__',
 '__dataframe__',
 '__dataframe_consortium_standard__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__finalize__',
 '__firstlineno__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__iand__',
 '__ifloordiv__',
 '__imod__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__invert__',
 '__ior__',
 '__ipow__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__ixor__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__

In [34]:
json_ou = response.json()
json_ou.keys()

dict_keys(['Error Message'])

In [30]:
json_ou['Meta Data']

{'1. Information': 'Daily Prices (open, high, low, close) and Volumes',
 '2. Symbol': 'TSLA',
 '3. Last Refreshed': '2025-01-08',
 '4. Output Size': 'Full size',
 '5. Time Zone': 'US/Eastern'}

In [31]:
json_ou['Time Series (Daily)']

{'2025-01-08': {'1. open': '392.9500',
  '2. high': '402.4999',
  '3. low': '387.4000',
  '4. close': '394.9400',
  '5. volume': '73038805'},
 '2025-01-07': {'1. open': '405.8300',
  '2. high': '414.3300',
  '3. low': '390.0000',
  '4. close': '394.3600',
  '5. volume': '75699525'},
 '2025-01-06': {'1. open': '423.2000',
  '2. high': '426.4300',
  '3. low': '401.7000',
  '4. close': '411.0500',
  '5. volume': '85516534'},
 '2025-01-03': {'1. open': '381.4800',
  '2. high': '411.8799',
  '3. low': '379.4500',
  '4. close': '410.4400',
  '5. volume': '95423329'},
 '2025-01-02': {'1. open': '390.1000',
  '2. high': '392.7299',
  '3. low': '373.0400',
  '4. close': '379.2800',
  '5. volume': '109710749'},
 '2024-12-31': {'1. open': '423.7900',
  '2. high': '427.9300',
  '3. low': '402.5400',
  '4. close': '403.8400',
  '5. volume': '76825121'},
 '2024-12-30': {'1. open': '419.4000',
  '2. high': '427.0000',
  '3. low': '415.7500',
  '4. close': '417.4100',
  '5. volume': '64941012'},
 '202

In [11]:
pd.DataFrame.from_dict(response, orient='index')

NameError: name 'pd' is not defined

In [4]:
from data_retrieval import DataRetrieval

DR = DataRetrieval(ticker=ticker)

hist_df = DR.Histrocal_prices()

In [5]:
hist_df.head()

Open      High     Low   Close      Volume
Date                                                    
2025-01-10  391.40  399.2800  377.29  394.74  62287333.0
2025-01-08  392.95  402.4999  387.40  394.94  73038805.0
2025-01-07  405.83  414.3300  390.00  394.36  75699525.0
2025-01-06  423.20  426.4300  401.70  411.05  85516534.0
2025-01-03  381.48  411.8799  379.45  410.44  95423329.0

In [13]:
hist_df.index = pd.to_datetime(hist_df.index)

#hist_df.index.name = "Dates"

hist_df.head()

Open      High       Low     Close    Volume
Dates                                                       
2025-01-10  391.4000  399.2800  377.2900  394.7400  62287333
2025-01-08  392.9500  402.4999  387.4000  394.9400  73038805
2025-01-07  405.8300  414.3300  390.0000  394.3600  75699525
2025-01-06  423.2000  426.4300  401.7000  411.0500  85516534
2025-01-03  381.4800  411.8799  379.4500  410.4400  95423329

In [9]:
cols = {col: col[3:].title() for col in hist_df.columns}

In [6]:
hist_df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 3658 entries, 2025-01-10 to 2010-06-29
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Open    3658 non-null   float64
 1   High    3658 non-null   float64
 2   Low     3658 non-null   float64
 3   Close   3658 non-null   float64
 4   Volume  3658 non-null   float64
dtypes: float64(5)
memory usage: 171.5 KB


In [10]:
cols

{'1. open': 'Open',
 '2. high': 'High',
 '3. low': 'Low',
 '4. close': 'Close',
 '5. volume': 'Volume'}

In [7]:
hist_df.index

DatetimeIndex(['2025-01-10', '2025-01-08', '2025-01-07', '2025-01-06',
               '2025-01-03', '2025-01-02', '2024-12-31', '2024-12-30',
               '2024-12-27', '2024-12-26',
               ...
               '2010-07-13', '2010-07-12', '2010-07-09', '2010-07-08',
               '2010-07-07', '2010-07-06', '2010-07-02', '2010-07-01',
               '2010-06-30', '2010-06-29'],
              dtype='datetime64[ns]', name='Date', length=3658, freq=None)

### Data retrival - Dividends

In [6]:
# importing the modules
from  data_retrieval import DataRetrieval

# Initializing
ticker = 'AAPL'
DR = DataRetrieval(ticker=ticker)
Div_hist = DR.Dividends_hist()


In [7]:
# Checking the attributes
dir(Div_hist)

['T',
 '_AXIS_LEN',
 '_AXIS_ORDERS',
 '_AXIS_TO_AXIS_NUMBER',
 '_HANDLED_TYPES',
 '__abs__',
 '__add__',
 '__and__',
 '__annotations__',
 '__array__',
 '__array_priority__',
 '__array_ufunc__',
 '__bool__',
 '__class__',
 '__column_consortium_standard__',
 '__contains__',
 '__copy__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__finalize__',
 '__firstlineno__',
 '__float__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__iand__',
 '__ifloordiv__',
 '__imod__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__int__',
 '__invert__',
 '__ior__',
 '__ipow__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__ixor__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__',
 '__or__',
 '__pandas_priority__',
 '__pos__',
 '__pow__',
 '_

Series([], Name: Dividends, dtype: float64)

In [12]:
Div_hist.json()

{'Information': 'Thank you for using Alpha Vantage! This is a premium endpoint. You may subscribe to any of the premium plans at https://www.alphavantage.co/premium/ to instantly unlock all premium endpoints'}

In [8]:
import yfinance as yf

ticker = yf.Ticker(ticker)
dividends = ticker.dividends
print(dividends)


Date
1987-05-11 00:00:00-04:00    0.000536
1987-08-10 00:00:00-04:00    0.000536
1987-11-17 00:00:00-05:00    0.000714
1988-02-12 00:00:00-05:00    0.000714
1988-05-16 00:00:00-04:00    0.000714
                               ...   
2023-11-10 00:00:00-05:00    0.240000
2024-02-09 00:00:00-05:00    0.240000
2024-05-10 00:00:00-04:00    0.250000
2024-08-12 00:00:00-04:00    0.250000
2024-11-08 00:00:00-05:00    0.250000
Name: Dividends, Length: 85, dtype: float64


In [15]:
dividends

Date
1987-05-11 00:00:00-04:00    0.000536
1987-08-10 00:00:00-04:00    0.000536
1987-11-17 00:00:00-05:00    0.000714
1988-02-12 00:00:00-05:00    0.000714
1988-05-16 00:00:00-04:00    0.000714
                               ...   
2023-11-10 00:00:00-05:00    0.240000
2024-02-09 00:00:00-05:00    0.240000
2024-05-10 00:00:00-04:00    0.250000
2024-08-12 00:00:00-04:00    0.250000
2024-11-08 00:00:00-05:00    0.250000
Name: Dividends, Length: 85, dtype: float64

In [3]:
from alpha_vantage.fundamentaldata import FundamentalData

fd = FundamentalData(key=settings.alpha_api_key, output_format='pandas')
data, _ = fd.get_company_overview(symbol="AAPL")
print(data["DividendPerShare"])


None    0.98
Name: DividendPerShare, dtype: object


In [4]:
data

Symbol     AssetType       Name  \
None   AAPL  Common Stock  Apple Inc   

                                            Description     CIK Exchange  \
None  Apple Inc. is an American multinational techno...  320193   NASDAQ   

     Currency Country      Sector              Industry  ... EVToRevenue  \
None      USD     USA  TECHNOLOGY  ELECTRONIC COMPUTERS  ...        9.26   

     EVToEBITDA  Beta 52WeekHigh 52WeekLow 50DayMovingAverage  \
None      26.89  1.24      260.1    163.49             238.71   

     200DayMovingAverage SharesOutstanding DividendDate ExDividendDate  
None              215.46       15115800000   2024-11-14     2024-11-08  

[1 rows x 52 columns]

In [5]:
import requests

api_key = settings.finnhub_api_key
symbol = "AAPL"
url = f"https://finnhub.io/api/v1/stock/dividend?symbol={symbol}&token={api_key}"
response = requests.get(url)
print(response.json())


{'error': "You don't have access to this resource."}


### Earnings

In [1]:
from data_retrieval import DataRetrieval
ticker = "TSLA"
DR = DataRetrieval(ticker=ticker)


earnings = DR.Earnings()

In [9]:
import pandas as pd
df = pd.DataFrame(earnings['quarterlyEarnings'])

In [22]:
df = (
    pd.DataFrame(earnings['annualEarnings'])
    .set_index("fiscalDateEnding")
    .astype(float)
)

In [23]:
df

reportedEPS
fiscalDateEnding             
2024-12-31             1.6900
2023-12-31             3.1300
2022-12-31             4.0700
2021-12-31             2.2600
2020-12-31             0.7500
2019-12-31             0.0700
2018-12-31            -0.0600
2017-12-31            -0.5700
2016-12-31            -0.1100
2015-12-31            -1.3000
2014-12-31             0.0300
2013-12-31             0.0500
2012-12-31            -0.2100
2011-12-31            -0.1600
2010-12-31            -0.0771

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15 entries, 2024-12-31 to 2010-12-31
Data columns (total 1 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   reportedEPS  15 non-null     float64
dtypes: float64(1)
memory usage: 240.0+ bytes


In [1]:
# Checking the earnings method
from data_retrieval import DataRetrieval
ticker = "TSLA"
DR = DataRetrieval(ticker=ticker)


q_earnings, a_earnigns = DR.Earnings()

In [2]:
q_earnings.head()


fiscalDateEnding reportedDate reportedEPS estimatedEPS surprise  \
0       2024-09-30   2024-10-23        0.72         0.58     0.14   
1       2024-06-30   2024-07-23        0.52         0.56    -0.04   
2       2024-03-31   2024-04-23        0.45         0.51    -0.06   
3       2023-12-31   2024-01-24        0.71         0.74    -0.03   
4       2023-09-30   2023-10-18        0.66         0.72    -0.06   

  surprisePercentage   reportTime  
0            24.1379  post-market  
1            -7.1429  post-market  
2           -11.7647  post-market  
3            -4.0541  post-market  
4            -8.3333  post-market

In [3]:
a_earnigns

reportedEPS
fiscalDateEnding             
2024-12-31             1.6900
2023-12-31             3.1300
2022-12-31             4.0700
2021-12-31             2.2600
2020-12-31             0.7500
2019-12-31             0.0700
2018-12-31            -0.0600
2017-12-31            -0.5700
2016-12-31            -0.1100
2015-12-31            -1.3000
2014-12-31             0.0300
2013-12-31             0.0500
2012-12-31            -0.2100
2011-12-31            -0.1600
2010-12-31            -0.0771

### Sector or Industry Data

In [1]:
from  data_retrieval import DataRetrieval
ticker = 'TSLA'
DR = DataRetrieval(ticker=ticker)



In [2]:
sect_indu = DR.sector_industry_data()

In [4]:
sect_indu.status_code

200

In [3]:
sect_indu

('MANUFACTURING', 'MOTOR VEHICLES & PASSENGER CAR BODIES')

In [ ]:
        sector_etfs = {
            "Technology": "XLK",
            "Financials": "XLF",
            "Healthcare": "XLV",
            "Energy": "XLE",
            "Consumer Discretionary": "XLY",
            "Consumer Staples": "XLP"
        }

In [1]:
import requests
from bs4 import BeautifulSoup

# URL for the ETFs by Sector page
url = "https://etfdb.com/etfs/by_sector/"

# Send a GET request to fetch the page
response = requests.get(url)
soup = BeautifulSoup(response.text, "html.parser")

# Example of scraping sector data and ETFs
sectors = soup.find_all("div", class_="etf-accordion")
sector_etf_map = {}

for sector in sectors:
    sector_name = sector.find("h2").text.strip()  # Extract sector name
    etfs = sector.find_all("a", class_="etfdb-link")  # Extract ETFs within each sector
    etf_list = [etf.text.strip() for etf in etfs]
    
    sector_etf_map[sector_name] = etf_list

# Print the sector-ETF mapping
print(sector_etf_map)


{}


In [ ]:
import requests
from bs4 import BeautifulSoup

# URL for the ETFs 
url = "https://etfdb.com/etfs/by_sector/"

# Send a GET request to  fethc  the page
response = requests.get(url)

response

TypeError: 'str' object is not callable

In [1]:
import requests
from bs4 import BeautifulSoup

# URL for the ETFs by sector page
url = "https://etfdb.com/etfs/sector/"

# Define headers to simulate a real browser request
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
}

# Send a GET request with the headers to fetch the page
response = requests.get(url, headers=headers)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    print("Request successful!")
    # Now parse the content with BeautifulSoup
    soup = BeautifulSoup(response.text, "html.parser")
    # Continue with your scraping logic...
else:
    print(f"Request failed with status code: {response.status_code}")


Request failed with status code: 403


### Final Testing of the class

In [2]:
from data_retrieval import DataRetrieval
import pandas as pd



In [5]:
tick = "TSLA"
data_retr = DataRetrieval(ticker=tick)

In [6]:
tsla_data = data_retr.historical_prices()

In [7]:
tsla_data.head()


Open      High      Low   Close      Volume
Date                                                     
2025-01-15  409.90  429.8000  405.661  428.22  79695189.0
2025-01-14  414.34  422.6400  394.540  396.36  84565022.0
2025-01-13  383.21  403.7900  380.070  403.31  67580494.0
2025-01-10  391.40  399.2800  377.290  394.74  62287333.0
2025-01-08  392.95  402.4999  387.400  394.94  73038805.0

In [3]:
# Testing Loop
import pandas as pd  # Ensure pandas is imported

# Ticker list with both valid and invalid entries
ticker_list = ['TSLA', 'TLSA', 'TSLAA', 'ELON', ' TSLA ']  # Well-informed test list

for ticker in ticker_list:
    print(f"Processing ticker: '{ticker}'")
    
    # Create an instance of DataRetrieval for the current ticker
    try:
        d_r = DataRetrieval(ticker=ticker)
        data = d_r.historical_prices()  # Fetch historical prices
        
        # Check if the returned data is a DataFrame
        if isinstance(data, pd.DataFrame):
            print(f"******* {ticker} **********")
            print(data.head())  # Display the first few rows of valid data
        else:
            print(f"Error: Ticker {ticker} returned a non-DataFrame result: {data}")
    
    # Handle unexpected errors during the data retrieval process
    except Exception as e:
        print(f"Exception occurred for ticker {ticker}: {e}")


Processing ticker: 'TSLA'
******* TSLA **********
              Open      High      Low   Close      Volume
Date                                                     
2025-01-15  409.90  429.8000  405.661  428.22  79695189.0
2025-01-14  414.34  422.6400  394.540  396.36  84565022.0
2025-01-13  383.21  403.7900  380.070  403.31  67580494.0
2025-01-10  391.40  399.2800  377.290  394.74  62287333.0
2025-01-08  392.95  402.4999  387.400  394.94  73038805.0
Processing ticker: 'TLSA'
******* TLSA **********
             Open    High     Low   Close     Volume
Date                                                
2025-01-15  0.728  0.8000  0.6800  0.7876   417253.0
2025-01-14  0.750  0.7649  0.7200  0.7250   134546.0
2025-01-13  0.848  0.8480  0.7200  0.7541   396289.0
2025-01-10  0.806  0.9500  0.8060  0.8400  2303389.0
2025-01-08  0.731  0.7360  0.7011  0.7360   116266.0
Processing ticker: 'TSLAA'
Exception occurred for ticker TSLAA: 'Error: Time Series (Daily) data not found in response. Che

In [1]:
import pandas as pd  # Ensure pandas is imported
from data_retrieval import DataRetrieval

# Enhanced ticker list with valid, invalid, and edge-case tickers
ticker_list = ['TSLA', 'AAPL', 'AMZN', 'MSFT', 'GOOGL', 'TLSA', 'TSLAA', 'ELON', 'BRK.B', 'SPY', ' TSLA ']

print("\n--- Testing Dividends History Retrieval ---\n")

for ticker in ticker_list:
    print(f"Processing ticker: '{ticker.strip()}'")  # Clean up any accidental whitespace in the ticker

    try:
        # Create an instance of DataRetrieval for the current ticker
        d_r = DataRetrieval(ticker=ticker.strip())  # Stripping whitespace for safety

        # Fetch the dividend history
        data = d_r.dividends_history()

        # Check if the returned data is a Series
        if isinstance(data, pd.Series):
            print(f"\n******* {ticker.strip()} **********")
            if not data.empty:
                print(data.head())  # Display the first few rows of valid data
            else:
                print("No dividend data available.")
        else:
            print(f"Error: Unexpected result type for ticker {ticker.strip()}: {type(data)}")

    except Exception as e:
        # Handle unexpected errors during the data retrieval process
        print(f"Exception occurred for ticker {ticker.strip()}: {e}")

print("\n--- End of Testing ---")



--- Testing Dividends History Retrieval ---

Processing ticker: 'TSLA'

******* TSLA **********
No dividend data available.
Processing ticker: 'AAPL'

******* AAPL **********
Date
1987-05-11 00:00:00-04:00    0.000536
1987-08-10 00:00:00-04:00    0.000536
1987-11-17 00:00:00-05:00    0.000714
1988-02-12 00:00:00-05:00    0.000714
1988-05-16 00:00:00-04:00    0.000714
Name: Date, dtype: float64
Processing ticker: 'AMZN'

******* AMZN **********
No dividend data available.
Processing ticker: 'MSFT'

******* MSFT **********
Date
2003-02-19 00:00:00-05:00    0.08
2003-10-15 00:00:00-04:00    0.16
2004-08-23 00:00:00-04:00    0.08
2004-11-15 00:00:00-05:00    3.08
2005-02-15 00:00:00-05:00    0.08
Name: Date, dtype: float64
Processing ticker: 'GOOGL'

******* GOOGL **********
Date
2024-06-10 00:00:00-04:00    0.2
2024-09-09 00:00:00-04:00    0.2
2024-12-09 00:00:00-05:00    0.2
Name: Date, dtype: float64
Processing ticker: 'TLSA'

******* TLSA **********
No dividend data available.
Process

$TSLAA: possibly delisted; no timezone found



******* TSLAA **********
No dividend data available.
Processing ticker: 'ELON'

******* ELON **********
No dividend data available.
Processing ticker: 'BRK.B'


$BRK.B: possibly delisted; no timezone found



******* BRK.B **********
No dividend data available.
Processing ticker: 'SPY'

******* SPY **********
Date
1993-03-19 00:00:00-05:00    0.213
1993-06-18 00:00:00-04:00    0.318
1993-09-17 00:00:00-04:00    0.286
1993-12-17 00:00:00-05:00    0.317
1994-03-18 00:00:00-05:00    0.271
Name: Date, dtype: float64
Processing ticker: 'TSLA'

******* TSLA **********
No dividend data available.

--- End of Testing ---


In [3]:
# Testing Loop for sector_industry_data
from data_retrieval import DataRetrieval
ticker_list = ['AAPL', 'TSLA', 'GOOGL', 'INVALID', 'META']

for ticker in ticker_list:
    print(f"\nProcessing ticker: {ticker}")
    
    try:
        d_r = DataRetrieval(ticker=ticker)
        details = d_r.sector_industry_data()  # Fetch sector/industry data
        
        print(f"Details for {ticker}: \n {details}")        
        #print(details.to_string(index=False))  # Print the DataFrame in a clean format
    
    except Exception as e:
        print(f"Error processing ticker {ticker}: {e}")



Processing ticker: AAPL
Details for AAPL: 
   Company Name Sector Industry
0          N/A    N/A      N/A

Processing ticker: TSLA
Details for TSLA: 
   Company Name Sector Industry
0          N/A    N/A      N/A

Processing ticker: GOOGL
Details for GOOGL: 
   Company Name Sector Industry
0          N/A    N/A      N/A

Processing ticker: INVALID
Details for INVALID: 
   Company Name Sector Industry
0          N/A    N/A      N/A

Processing ticker: META
Details for META: 
   Company Name Sector Industry
0          N/A    N/A      N/A


In [6]:
my_dict = {'name': 'Omera', 'age': 25}

# Using .get()
name = my_dict.get('Name', 'Error')  # Returns 'Omera'
print(name)

# Accessing a missing key
hobby = my_dict.get('hobby')  # Returns None (no KeyError)
print(hobby)


Error
None


In [1]:
from data_retrieval import DataRetrieval
tic= "APLL"
d_r = DataRetrieval(ticker=tic)
details = d_r.sector_industry_data()
details

<Response [200]>

In [2]:
details.json()

{'Information': 'Thank you for using Alpha Vantage! Our standard API rate limit is 25 requests per day. Please subscribe to any of the premium plans at https://www.alphavantage.co/premium/ to instantly remove all daily rate limits.'}

In [2]:
from config import settings

SettingsError: error parsing value for field "alpha_api_keys" from source "DotEnvSettingsSource"

In [5]:
dir(settings.json())

C:\Users\USER\AppData\Local\Temp\ipykernel_18092\713758585.py:1: PydanticDeprecatedSince20: The `json` method is deprecated; use `model_dump_json` instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.10/migration/
  dir(settings.json())


['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

In [6]:
settings.json()

C:\Users\USER\AppData\Local\Temp\ipykernel_18092\2485054537.py:1: PydanticDeprecatedSince20: The `json` method is deprecated; use `model_dump_json` instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.10/migration/
  settings.json()


'{"alpha_api_key":"a91a979443754148c158f79e886355890636122f108d85cf50acbb11e347a87d39538a0c4ee017a7b439d3c7d9c6468ff851ba568aef40e1b42826ce5779d6370544af8680b2175b1f5fae18bfe2e2f6ffce24355a51a74b95993b1fbafcbc2e07cbcee752948424a4dfb2be2f2071a78cf0c671c3dfe3dc97cbe09725ea7443","finnhub_api_key":"cu2fta9r01ql7sc7c0pgcu2fta9r01ql7sc7c0q0"}'

In [1]:
from config import settings

print("Alpha API Keys:", settings.alpha_api_keys)
print("Proxy List:", settings.proxy_list)


SettingsError: error parsing value for field "alpha_api_keys" from source "DotEnvSettingsSource"

In [1]:
from config import settings

print("Alpha API Keys:", settings.alpha_api_keys)
print("Proxy List:", settings.proxy_list)


Alpha API Keys: ['a91a979443754148c158f79e886355890636122f108d85cf50acbb11e347a87d39538a0c4ee017a7b439d3c7d9c6468ff851ba568aef40e1b42826ce5779d6370544af8680b2175b1f5fae18bfe2e2f6ffce24355a51a74b95993b1fbafcbc2e07cbcee752948424a4dfb2be2f2071a78cf0c671c3dfe3dc97cbe09725ea7443', '5414505df052d9c3ac11298120691bd4d781a589e1930ddb4423783e7f1b6592a27a63ac03f0bf51c9345ed502f9b418ee8b54edcd86fe62f97f6fd70539f36d0ab962579fe39d81f97ea82bb6ee850de8bee15850a9b2afaf1456cd97fa0440e56586cf3a3c4cb699ebf002369e1575e221d29d585f2ec316756dd8b2f4d383', 'XS0UJ1Z2PIJSIHP9']
Proxy List: ['http://proxy1:port', 'http://proxy2:port', 'http://proxy3:port']


In [2]:
import os
from dotenv import load_dotenv

load_dotenv()
print(os.getenv("ALPHA_API_KEYS"))


a91a979443754148c158f79e886355890636122f108d85cf50acbb11e347a87d39538a0c4ee017a7b439d3c7d9c6468ff851ba568aef40e1b42826ce5779d6370544af8680b2175b1f5fae18bfe2e2f6ffce24355a51a74b95993b1fbafcbc2e07cbcee752948424a4dfb2be2f2071a78cf0c671c3dfe3dc97cbe09725ea7443,5414505df052d9c3ac11298120691bd4d781a589e1930ddb4423783e7f1b6592a27a63ac03f0bf51c9345ed502f9b418ee8b54edcd86fe62f97f6fd70539f36d0ab962579fe39d81f97ea82bb6ee850de8bee15850a9b2afaf1456cd97fa0440e56586cf3a3c4cb699ebf002369e1575e221d29d585f2ec316756dd8b2f4d383,XS0UJ1Z2PIJSIHP9


### Proxy Spinning 

Here's a detailed guide addressing your concerns about proxies and their use in Python applications:

---

#### **Proxy Basics**

1. **What are proxies, and how do they work in API requests?**
   - Proxies act as intermediaries between your application and the target server. They route your requests through a different IP address, masking your actual location.
   - When making an API request via a proxy, the target server sees the proxy's IP instead of your own, which helps bypass rate limits or geographic restrictions.

2. **Differences between Free Proxies, Paid Proxies, and Rotating Proxies:**
   - **Free Proxies:** 
     - Free of cost, but often unreliable, slow, and potentially insecure.
     - Many are shared, which can lead to rate limits and bans.
   - **Paid Proxies:** 
     - Offer dedicated or semi-dedicated IPs with better speed, reliability, and support.
     - Less likely to be flagged or blacklisted due to misuse.
   - **Rotating Proxies:** 
     - Automatically change IPs at regular intervals or after every request.
     - Ideal for web scraping or API interactions to avoid detection and bans.

3. **Reliability of Free Proxies Compared to Paid Services:**
   - Free proxies are generally less reliable, with frequent downtimes and lower speeds.
   - Paid services offer higher reliability, anonymity, and performance, making them more suitable for production-grade applications.

---

#### **Testing Proxies**

1. **Verifying Proxy Status:**
   - Use tools like `requests`, `httpx`, or `aiohttp` to test if a proxy is active.
   - Validate by sending a request to a test URL like `httpbin.org/ip`.

2. **Tools, Libraries, or Methods:**
   - **Python Libraries:**
     - `requests`: For basic proxy testing.
     - `httpx`: Supports asynchronous requests for faster validation.
     - `pycurl`: For detailed control over HTTP requests.
   - **Metrics to Test:**
     - Response time (latency).
     - Status codes (200 for success).
     - Content validation (e.g., returned IP matches proxy IP).

3. **Handling Common Issues:**
   - **Blocked Proxies:** Switch to a new proxy or use a rotating proxy service.
   - **Slow Response Times:** Set timeout limits and filter out slow proxies.
   - **Incorrect Configurations:** Double-check the format (`http://user:pass@ip:port`).

---

#### **Proxy Spinning Implementation**

1. **Dynamic Proxy Rotation:**
   - Maintain a pool of proxies and rotate them in a round-robin fashion or randomly.
   - Use rotating proxy services (e.g., BrightData, ScraperAPI) for automatic IP changes.

2. **Fallback Mechanism:**
   - Implement retries with a different proxy if one fails.
   - Log failures to avoid reusing problematic proxies.

---

-

#### **Troubleshooting Tips**

1. **Identifying Blocked Proxies:**
   - Look for:
     - HTTP 403 Forbidden or similar status codes.
     - CAPTCHA challenges in responses.
   - Solutions:
     - Rotate the proxy.
     - Add headers or mimic browser-like behavior.

2. **Improving Success Rates:**
   - Use high-quality, paid, or rotating proxies.
   - Adjust request intervals to avoid triggering rate limits.
   - Add random delays between requests.

---

#### **Recommendations & Pitfalls to Avoid**

- **Use HTTPS:** Ensure proxies support HTTPS for secure API communication.
- **Monitor Proxy Health:** Regularly test proxies to remove inactive ones.
- **Avoid Overloading a Single Proxy:** Spread requests evenly to reduce bans.
- **Handle Exceptions Gracefully:** Implement robust error handling to recover from failures.

By following these practices, you can effectively implement and manage proxy spinning for your Python application.

In [3]:
import requests
import time

# List of Proxies
proxies = [
    "http://scraperapi:650e2639b05cfff8c62185e0723baca0@proxy-server.scraperapi.com:8001"

]

# Test URL
test_url = "http://httpbin.org/ip"

# Function to test proxies
def test_proxies(proxy_list):
    valid_proxies = []
    print(f"{'Proxy':<40} {'Status':<10} {'Response Time (ms)':<20}")
    print("-" * 70)

    for proxy in proxy_list:
        proxy_dict = {"http": proxy, "https": proxy}
        try:
            start = time.time()
            response = requests.get(test_url, proxies=proxy_dict, timeout=5)
            end = time.time()
            response_time = (end - start) * 1000  # Convert to milliseconds

            if response.status_code == 200:
                print(f"{proxy:<40} {'Success':<10} {response_time:<20.2f}")
                valid_proxies.append(proxy)
            else:
                print(f"{proxy:<40} {'Fail':<10} {'-':<20}")
        except Exception as e:
            print(f"{proxy:<40} {'Fail':<10} {'-':<20}")

    print("\nUsable Proxies:")
    for valid_proxy in valid_proxies:
        print(valid_proxy)

    return valid_proxies

# Test the proxies
usable_proxies = test_proxies(proxies)

Proxy                                    Status     Response Time (ms)  
----------------------------------------------------------------------
http://scraperapi:650e2639b05cfff8c62185e0723baca0@proxy-server.scraperapi.com:8001 Fail       -                   

Usable Proxies:


In [5]:
import requests
proxies = {
"http": "http://scraperapi.render=true:650e2639b05cfff8c62185e0723baca0@proxy-server.scraperapi.com:8001"
}
r = requests.get('http://httpbin.org/ip', proxies=proxies, verify=False)
print(r.text)

ProxyError: HTTPConnectionPool(host='proxy-server.scraperapi.com', port=8001): Max retries exceeded with url: http://httpbin.org/ip (Caused by ProxyError('Unable to connect to proxy', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None)))

In [44]:
import requests

def test_proxy(proxy):
    url= "https://www.youtube.com/"
    proxies = {
        "http": proxy,
        "https": proxy
    }
    try:
        response = requests.get(url, proxies=proxies)
        if response.status_code == 200:
            print(f"PRoxy Wking: {proxy}")
            print(f"Response: {response.json()}")
        else:
            print(f"proxy failed with status code: {response.status_code}")
    except requests.exceptions.RequestException as e:
        print(f"Error with  proxy {proxy}: {e}")
proxy_list = [
    "http://94.68.245.147:8080",
    "http://38.156.75.235:8080",
    "http://4.241.132.44:3128",
    "http://107.181.168.145:4145",

]

for proxy in proxy_list:
    test_proxy(proxy)

PRoxy Wking: http://94.68.245.147:8080
Error with  proxy http://94.68.245.147:8080: Expecting value: line 1 column 1 (char 0)
Error with  proxy http://38.156.75.235:8080: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))
Error with  proxy http://4.241.132.44:3128: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))
Error with  proxy http://107.181.168.145:4145: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))


In [7]:
proxy_list = [
    "185.105.102.189:80",
    "124.122.114.109:8080",
    "4.241.132.44:3128",
    "107.181.168.145:4145",

]

In [8]:
for i in proxy_list:
    print(i)

185.105.102.189:80
124.122.114.109:8080
4.241.132.44:3128
107.181.168.145:4145


In [40]:
url= "https://www.the-star.co.ke/news/2025-01-16-nyaga-named-new-police-spokesperson-in-major-nps-reshuffle"
proxy = {"http" :"http://124.122.114.109:8080"}

res = requests.get(url=url, proxies=proxy)

In [42]:
res.status_code

200

In [ ]:
import requests

def test_proxy(proxy):
    url = "https://www.alphavantage.co/query?function=TIME_SERIES_DAILY&symbol=AMBUJACEM.BSE&apikey=a91a979443754148c158f79e886355890636122f108d85cf50acbb11e347a87d39538a0c4ee017a7b439d3c7d9c6468ff851ba568aef40e1b42826ce5779d6370544af8680b2175b1f5fae18bfe2e2f6ffce24355a51a74b95993b1fbafcbc2e07cbcee752948424a4dfb2be2f2071a78cf0c671c3dfe3dc97cbe09725ea7443"
    proxies = {
        "http": proxy,
        "https": proxy
    }
    
    with requests.Session() as session:
        session.proxies.update(proxies)
        try:
            response = session.get(url, timeout=10)
            if response.status_code == 200:
                print(f"Proxy Working: {proxy}")
                print(f"Response: {response.json()}")
            else:
                print(f"Proxy failed with status code: {response.status_code}")
        except requests.exceptions.RequestException as e:
            print(f"Error with proxy {proxy}: {e}")

proxy_list = [
    "http://185.105.102.189:80",
    "http://124.122.114.109:8080",
    "http://4.241.132.44:3128",
    "http://107.181.168.145:4145",
]

for proxy in proxy_list:
    test_proxy(proxy)


In [21]:
import random
import requests

def test_proxy(proxy):
    url = "https://example.com/"
    proxies = {
        "http": proxy,
        "https": proxy
    }
    session = requests.Session()
    session.proxies.update(proxies)
    
    try:
        # Disable SSL verification
        response = session.get(url, timeout=5, verify=False)
        
        if response.status_code == 200:
            print(f"Proxy Working: {proxy}")
            print(f"Response: {response.json()}")
        else:
            print(f"Proxy failed with status code: {response.status_code}")
    except requests.exceptions.RequestException as e:
        print(f"Error with proxy {proxy}: {e}")


def test_random_proxies(proxy_list, n_tests):
    proxy_list = proxy_list.tolist()  # Convert Series to list
    random.shuffle(proxy_list)  # Shuffle proxies list
    selected_proxies = random.sample(proxy_list, n_tests)  # Pick random proxies
    
    for proxy in selected_proxies:
        test_proxy(proxy)




In [22]:
import pandas as pd
proxy_list_raw = pd.read_csv(r"C:\Users\USER\OneDrive\Projects_2025\Intelligent-Trading-solution\plainproxies_com_free_proxy_list_all.txt", header=None).rename(columns={0: "proxy"})
# formatting the proxies 
proxy_list_raw['updated_proxy'] = proxy_list_raw['proxy'].apply(lambda x: f"http://{x}")
proxy_list = proxy_list_raw['updated_proxy'].squeeze()

In [19]:
proxy_list.info()

<class 'pandas.core.series.Series'>
RangeIndex: 587 entries, 0 to 586
Series name: updated_proxy
Non-Null Count  Dtype 
--------------  ----- 
587 non-null    object
dtypes: object(1)
memory usage: 4.7+ KB


In [23]:
n_tests = 5  # Specify the number of proxies to test
test_random_proxies(proxy_list, n_tests)

c:\Users\USER\anaconda3\envs\ITS\Lib\site-packages\urllib3\connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host '147.75.34.92'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(


Error with proxy http://147.75.34.92:9443: HTTPSConnectionPool(host='example.com', port=443): Read timed out. (read timeout=5)


c:\Users\USER\anaconda3\envs\ITS\Lib\site-packages\urllib3\connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host '13.38.153.36'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(


Proxy Working: http://13.38.153.36:80
Error with proxy http://13.38.153.36:80: Expecting value: line 1 column 1 (char 0)
Error with proxy http://128.140.113.110:8081: HTTPSConnectionPool(host='example.com', port=443): Max retries exceeded with url: / (Caused by ProxyError('Unable to connect to proxy', OSError('Tunnel connection failed: 403 Forbidden')))


c:\Users\USER\anaconda3\envs\ITS\Lib\site-packages\urllib3\connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host '35.79.120.242'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(


Proxy Working: http://35.79.120.242:3128
Error with proxy http://35.79.120.242:3128: Expecting value: line 1 column 1 (char 0)
Error with proxy http://216.229.112.25:8080: HTTPSConnectionPool(host='example.com', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, '[SSL: UNEXPECTED_EOF_WHILE_READING] EOF occurred in violation of protocol (_ssl.c:1020)')))
